File: generated_code/python/solution_124_3.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        """
        Algorithm: Post-order Traversal (DFS)
        
        For each node, we calculate the maximum path sum that can extend upwards 
        to its parent. This is determined by the node's value plus the maximum 
        of its left or right subtree contributions (only if they are positive).
        
        Simultaneously, we track the global maximum path sum where the current 
        node acts as the highest point (the "root") of the path, combining 
        both left and right subtree contributions.
        
        Time Complexity: O(N) - Each node is visited exactly once.
        Space Complexity: O(H) - Max recursion stack depth proportional to tree height.
        """
        # Initialize global maximum with negative infinity
        self.max_sum = float('-inf')
        
        def get_max_gain(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            
            # Recursively get the max path sum from left and right subtrees.
            # If a subtree contribution is negative, we drop it (take 0).
            left_gain = max(get_max_gain(node.left), 0)
            right_gain = max(get_max_gain(node.right), 0)
            
            # Price of a new path with the current node as the highest point (turnaround point)
            current_path_sum = node.val + left_gain + right_gain
            
            # Update the global maximum path sum found so far
            self.max_sum = max(self.max_sum, current_path_sum)
            
            # Return the max gain the parent node can get by extending through this node
            return node.val + max(left_gain, right_gain)
        
        get_max_gain(root)
        return self.max_sum

File: generated_code/cpp/solution_124_3.cpp
#include <algorithm>
#include <climits>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 * int val;
 * TreeNode *left;
 * TreeNode *right;
 * TreeNode() : val(0), left(nullptr), right(nullptr) {}
 * TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int global_max_sum;

    // Helper function that returns the maximum path sum starting from the current node
    // and extending down into one of its subtrees.
    int calculateMaxGain(TreeNode* node) {
        if (node == nullptr) {
            return 0;
        }

        // Recursively find the max gain from left and right subtrees.
        // If the path sum from a subtree is negative, we drop it (take 0).
        int left_gain = std::max(calculateMaxGain(node->left), 0);
        int right_gain = std::max(calculateMaxGain(node->right), 0);

        // Price of a new path with the current node as the highest bridge node
        int current_path_sum = node->val + left_gain + right_gain;

        // Update the global maximum path sum found so far
        global_max_sum = std::max(global_max_sum, current_path_sum);

        // For the parent call, a path can only extend through one child
        return node->val + std::max(left_gain, right_gain);
    }

public:
    int maxPathSum(TreeNode* root) {
        // Initialize global maximum with the smallest possible integer to handle strictly negative nodes
        global_max_sum = INT_MIN;
        calculateMaxGain(root);
        return global_max_sum;
    }
};